File: typescript/src/generator.ts
/**
 * ORT Generator for TypeScript
 */

import { OrtValue, OrtValueType } from './value';

export function generateOrt(value: OrtValue): string {
    if (value.isObject()) {
        const obj = value.asObject();
        if (obj) {
            const objNative = value.toNative() as { [key: string]: any };

            // Check if this is a multi-key object
            if (Object.keys(objNative).length > 1 || Object.keys(objNative).length === 0) {
                return generateMultiObject(objNative);
            } else if (Object.keys(objNative).length === 1) {
                // Single key - might be a named array
                const [key, val] = Object.entries(objNative)[0];
                const valOrt = new OrtValue(val);
                if (valOrt.isArray()) {
                    const arr = valOrt.asArray();
                    if (!arr || arr.length === 0) {
                        return `${key}:\n[]`;
                    } else {
                        const arrList = arr.map(v => v.toNative());
                        if (isUniformObjectArray(arrList)) {
                            return generateObjectArray(key, arrList);
                        } else {
                            return generateSimpleArray(key, arrList);
                        }
                    }
                } else {
                    // Single key with non-array value
                    return `${key}:\n${generateValue(val, false)}`;
                }
            }
        }
        return '';
    } else if (value.isArray()) {
        const arr = value.asArray();
        if (arr) {
            const arrList = arr.map(v => v.toNative());
            // Top-level array
            if (isUniformObjectArray(arrList)) {
                return generateTopLevelObjectArray(arrList);
            } else {
                return `:${generateArrayContent(arrList, false)}`;
            }
        }
        return ':[]';
    } else {
        return generateValue(value.toNative(), false);
    }
}

function generateMultiObject(obj: { [key: string]: any }): string {
    const result: string[] = [];
    const entries = Object.entries(obj);

    for (let i = 0; i < entries.length; i++) {
        const [key, val] = entries[i];
        const valOrt = new OrtValue(val);
        if (valOrt.isArray()) {
            const arr = valOrt.asArray();
            if (!arr || arr.length === 0) {
                result.push(`${key}:\n[]`);
            } else {
                const arrList = arr.map(v => v.toNative());
                if (isUniformObjectArray(arrList)) {
                    result.push(generateObjectArray(key, arrList).trimEnd());
                } else {
                    result.push(generateSimpleArray(key, arrList).trimEnd());
                }
            }
        } else {
            result.push(`${key}:\n${generateValue(val, false)}`);
        }

        if (i < entries.length - 1) {
            result.push('\n\n');
        } else {
            result.push('\n');
        }
    }

    return result.join('');
}

function isUniformObjectArray(arr: any[]): boolean {
    if (arr.length === 0) {
        return false;
    }

    // Check if all elements are objects with the same keys
    if (typeof arr[0] !== 'object' || arr[0] === null || Array.isArray(arr[0])) {
        return false;
    }

    const firstKeys = Object.keys(arr[0]).sort();

    for (let i = 1; i < arr.length; i++) {
        const item = arr[i];
        if (typeof item !== 'object' || item === null || Array.isArray(item)) {
            return false;
        }
        const keys = Object.keys(item).sort();
        if (JSON.stringify(keys) !== JSON.stringify(firstKeys)) {
            return false;
        }
    }

    return true;
}

function generateObjectArray(key: string, arr: { [key: string]: any }[]): string {
    if (arr.length === 0) {
        return `${key}:\n[]`;
    }

    const first = arr[0];
    const keys = Object.keys(first);
    const header = generateHeader(keys, first);

    const result: string[] = [`${key}:${header}`];

    for (const item of arr) {
        const values = keys.map(k => generateObjectFieldValue(item[k], keys, k, item));
        result.push(values.join(','));
    }

    return result.join('\n');
}

function generateTopLevelObjectArray(arr: { [key: string]: any }[]): string {
    if (arr.length === 0) {
        return ':[]';
    }

    const first = arr[0];
    const keys = Object.keys(first);
    const header = generateHeader(keys, first);

    const result: string[] = [`:${header}`];

    for (const item of arr) {
        const values = keys.map(k => generateObjectFieldValue(item[k], keys, k, item));
        result.push(values.join(','));
    }

    return result.join('\n');
}

function generateHeader(keys: string[], firstObj: { [key: string]: any }): string {
    const headerParts: string[] = [];

    for (const k of keys) {
        const value = firstObj[k];
        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
            // Generate nested field
            const nestedKeys = Object.keys(value);
            const nestedHeader = generateHeaderFields(nestedKeys, value);
            headerParts.push(`${k}(${nestedHeader})`);
        } else {
            headerParts.push(k);
        }
    }

    return headerParts.join(',') + ':';
}

function generateHeaderFields(keys: string[], obj: { [key: string]: any }): string {
    const headerParts: string[] = [];

    for (const k of keys) {
        const value = obj[k];
        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
            // Recursively generate nested field
            const nestedKeys = Object.keys(value);
            const nestedHeader = generateHeaderFields(nestedKeys, value);
            headerParts.push(`${k}(${nestedHeader})`);
        } else {
            headerParts.push(k);
        }
    }

    return headerParts.join(',');
}

function generateObjectFieldValue(
    value: any,
    keys: string[],
    currentKey: string,
    parent: { [key: string]: any }
): string {
    if (value === null || value === undefined) {
        return '';
    } else if (typeof value === 'object' && !Array.isArray(value)) {
        if (Object.keys(value).length === 0) {
            return '()';
        } else {
            // Nested object - output values only (keys are in header)
            const nestedKeys = Object.keys(value);
            const values = nestedKeys.map(k => generateObjectFieldValue(value[k], nestedKeys, k, value));
            return `(${values.join(',')})`;
        }
    } else if (Array.isArray(value)) {
        if (value.length === 0) {
            return '[]';
        } else {
            return `[${generateArrayContent(value, true)}]`;
        }
    } else {
        return generateValue(value, true);
    }
}

function generateSimpleArray(key: string, arr: any[]): string {
    return `${key}:\n${generateArrayContent(arr, false)}`;
}

function generateArrayContent(arr: any[], inline: boolean): string {
    if (arr.length === 0) {
        return '[]';
    }

    const values = arr.map(v => generateValue(v, inline));

    if (inline) {
        return values.join(',');
    } else {
        return `[${values.join(',')}]`;
    }
}

function generateValue(value: any, inline: boolean): string {
    if (value === null || value === undefined) {
        return '';
    } else if (typeof value === 'boolean') {
        return value ? 'true' : 'false';
    } else if (typeof value === 'number') {
        // Format number nicely (remove .0 for whole numbers)
        if (Number.isInteger(value)) {
            return String(value);
        }
        return String(value);
    } else if (typeof value === 'string') {
        return escape(value);
    } else if (Array.isArray(value)) {
        if (value.length === 0) {
            return '[]';
        } else {
            return `[${generateArrayContent(value, true)}]`;
        }
    } else if (typeof value === 'object') {
        if (Object.keys(value).length === 0) {
            return '()';
        } else {
            return generateInlineObject(value);
        }
    } else {
        return String(value);
    }
}

function generateInlineObject(obj: { [key: string]: any }): string {
    const pairs = Object.entries(obj).map(([k, v]) => `${k}:${generateValue(v, true)}`);
    return `(${pairs.join(',')})`;
}

function escape(s: string): string {
    const result: string[] = [];

    for (const ch of s) {
        if (ch === '(') {
            result.push('\\(');
        } else if (ch === ')') {
            result.push('\\)');
        } else if (ch === '[') {
            result.push('\\[');
        } else if (ch === ']') {
            result.push('\\]');
        } else if (ch === ',') {
            result.push('\\,');
        } else if (ch === '\\') {
            result.push('\\\\');
        } else if (ch === '\n') {
            result.push('\\n');
        } else if (ch === '\t') {
            result.push('\\t');
        } else if (ch === '\r') {
            result.push('\\r');
        } else {
            result.push(ch);
        }
    }

    return result.join('');
}

File: rust/generator.rs
use crate::ort_value::OrtValue;
use std::collections::HashMap;

pub fn generate_ort(value: &OrtValue) -> String {
    match value {
        OrtValue::Object(obj) => {
            // Check if this is a multi-key object
            if obj.len() > 1 || obj.is_empty() {
                generate_multi_object(obj)
            } else if obj.len() == 1 {
                // Single key - might be a named array
                let (key, val) = obj.iter().next().unwrap();
                if let OrtValue::Array(arr) = val {
                    if arr.is_empty() {
                        format!("{}:\n[]", key)
                    } else if is_uniform_object_array(arr) {
                        generate_object_array(key, arr)
                    } else {
                        generate_simple_array(key, arr)
                    }
                } else {
                    // Single key with non-array value
                    format!("{}:\n{}", key, generate_value(val, false))
                }
            } else {
                String::new()
            }
        }
        OrtValue::Array(arr) => {
            // Top-level array
            if is_uniform_object_array(arr) {
                generate_top_level_object_array(arr)
            } else {
                format!(":{}", generate_array_content(arr, false))
            }
        }
        _ => generate_value(value, false),
    }
}

fn generate_multi_object(obj: &HashMap<String, OrtValue>) -> String {
    let mut result = Vec::new();
    let mut entries: Vec<_> = obj.iter().collect();
    entries.sort_by(|a, b| a.0.cmp(b.0));

    for (i, (key, val)) in entries.iter().enumerate() {
        if let OrtValue::Array(arr) = val {
            if is_uniform_object_array(arr) {
                result.push(generate_object_array(key, arr).trim_end().to_string());
            } else {
                result.push(generate_simple_array(key, arr).trim_end().to_string());
            }
        } else {
            result.push(format!("{}:\n{}", key, generate_value(val, false)));
        }

        if i < entries.len() - 1 {
            result.push("\n\n".to_string());
        } else {
            result.push("\n".to_string());
        }
    }

    result.join("")
}

fn is_uniform_object_array(arr: &[OrtValue]) -> bool {
    if arr.is_empty() {
        return false;
    }

    // Check if all elements are objects with the same keys
    let first_keys = match &arr[0] {
        OrtValue::Object(obj) => {
            let mut keys: Vec<_> = obj.keys().collect();
            keys.sort();
            keys
        }
        _ => return false,
    };

    for item in arr.iter().skip(1) {
        match item {
            OrtValue::Object(obj) => {
                let mut keys: Vec<_> = obj.keys().collect();
                keys.sort();
                if keys != first_keys {
                    return false;
                }
            }
            _ => return false,
        }
    }

    true
}

fn generate_object_array(key: &str, arr: &[OrtValue]) -> String {
    if arr.is_empty() {
        return format!("{}:\n[]", key);
    }

    let first = &arr[0];
    if let OrtValue::Object(obj) = first {
        let keys: Vec<_> = obj.keys().cloned().collect();
        let header = generate_header(&keys, obj);

        let mut result = vec![format!("{}:{}", key, header)];

        for item in arr {
            if let OrtValue::Object(obj) = item {
                let values: Vec<String> = keys
                    .iter()
                    .map(|k| generate_object_field_value(obj.get(k).unwrap_or(&OrtValue::Null), &keys, k, obj))
                    .collect();
                result.push(values.join(","));
            }
        }

        result.join("\n")
    } else {
        format!("{}:\n{}", key, generate_array_content(arr, false))
    }
}

fn generate_top_level_object_array(arr: &[OrtValue]) -> String {
    if arr.is_empty() {
        return ":[]".to_string();
    }

    let first = &arr[0];
    if let OrtValue::Object(obj) = first {
        let keys: Vec<_> = obj.keys().cloned().collect();
        let header = generate_header(&keys, obj);

        let mut result = vec![format!(":{}", header)];

        for item in arr {
            if let OrtValue::Object(obj) = item {
                let values: Vec<String> = keys
                    .iter()
                    .map(|k| generate_object_field_value(obj.get(k).unwrap_or(&OrtValue::Null), &keys, k, obj))
                    .collect();
                result.push(values.join(","));
            }
        }

        result.join("\n")
    } else {
        format!(":{}", generate_array_content(arr, false))
    }
}

fn generate_header(keys: &[String], first_obj: &HashMap<String, OrtValue>) -> String {
    keys.iter()
        .map(|k| {
            if let Some(value) = first_obj.get(k) {
                match value {
                    OrtValue::Object(nested_obj) => {
                        // Generate nested field
                        let nested_keys: Vec<_> = nested_obj.keys().cloned().collect();
                        let nested_header = generate_header_fields(&nested_keys, nested_obj);
                        format!("{}({})", k, nested_header)
                    }
                    _ => k.clone(),
                }
            } else {
                k.clone()
            }
        })
        .collect::<Vec<_>>()
        .join(",") + ":"
}

fn generate_header_fields(keys: &[String], obj: &HashMap<String, OrtValue>) -> String {
    keys.iter()
        .map(|k| {
            if let Some(value) = obj.get(k) {
                match value {
                    OrtValue::Object(nested_obj) => {
                        // Recursively generate nested field
                        let nested_keys: Vec<_> = nested_obj.keys().cloned().collect();
                        let nested_header = generate_header_fields(&nested_keys, nested_obj);
                        format!("{}({})", k, nested_header)
                    }
                    _ => k.clone(),
                }
            } else {
                k.clone()
            }
        })
        .collect::<Vec<_>>()
        .join(",")
}

fn generate_object_field_value(
    value: &OrtValue,
    _keys: &[String],
    _current_key: &str,
    _parent: &HashMap<String, OrtValue>,
) -> String {
    match value {
        OrtValue::Null => String::new(),
        OrtValue::Object(obj) => {
            if obj.is_empty() {
                "()".to_string()
            } else {
                // Nested object - output values only (keys are in header)
                let nested_keys: Vec<_> = obj.keys().cloned().collect();
                let values: Vec<String> = nested_keys
                    .iter()
                    .map(|k| {
                        generate_object_field_value(
                            obj.get(k).unwrap_or(&OrtValue::Null),
                            &nested_keys,
                            k,
                            obj,
                        )
                    })
                    .collect();
                format!("({})", values.join(","))
            }
        }
        OrtValue::Array(arr) => {
            if arr.is_empty() {
                "[]".to_string()
            } else {
                format!("[{}]", generate_array_content(arr, true))
            }
        }
        _ => generate_value(value, true),
    }
}

fn generate_simple_array(key: &str, arr: &[OrtValue]) -> String {
    format!("{}:\n{}", key, generate_array_content(arr, false))
}

fn generate_array_content(arr: &[OrtValue], inline: bool) -> String {
    if arr.is_empty() {
        return "[]".to_string();
    }

    let values: Vec<String> = arr.iter().map(|v| generate_value(v, inline)).collect();

    if inline {
        values.join(",")
    } else {
        format!("[{}]", values.join(","))
    }
}

fn generate_value(value: &OrtValue, _inline: bool) -> String {
    match value {
        OrtValue::Null => String::new(),
        OrtValue::Bool(b) => b.to_string(),
        OrtValue::Number(n) => n.to_string(),
        OrtValue::String(s) => escape(s),
        OrtValue::Array(arr) => {
            if arr.is_empty() {
                "[]".to_string()
            } else {
                format!("[{}]", generate_array_content(arr, true))
            }
        }
        OrtValue::Object(obj) => {
            if obj.is_empty() {
                "()".to_string()
            } else {
                generate_inline_object(obj)
            }
        }
    }
}

fn generate_inline_object(obj: &HashMap<String, OrtValue>) -> String {
    let pairs: Vec<String> = obj
        .iter()
        .map(|(k, v)| format!("{}:{}", k, generate_value(v, true)))
        .collect();
    format!("({})", pairs.join(","))
}

fn escape(s: &str) -> String {
    let mut result = String::new();

    for ch in s.chars() {
        match ch {
            '(' => result.push_str("\\("),
            ')' => result.push_str("\\)"),
            '[' => result.push_str("\\["),
            ']' => result.push_str("\\]"),
            ',' => result.push_str("\\,"),
            '\\' => result.push_str("\\\\"),
            '\n' => result.push_str("\\n"),
            '\t' => result.push_str("\\t"),
            '\r' => result.push_str("\\r"),
            _ => result.push(ch),
        }
    }

    result
}
